# Assignment - 4

### Theory questions

Q 1. What is the difference between a function and a method in python?

1. Function:

✲ A function is a block of code that performs a specific task.

✲ It can be defined using the def keyword.

✲ Functions are generally independent of any class or object. 

✲ They can exist on their own and can be called directly without   needing an object.

✲ Functions can be called by their name and passed arguments to    work with.
Example-

In [3]:
def greet(name):
    return f"hello , {name}"

greet('Tushar')    

'hello , Tushar'

2. Method:

✲ A method is a function that is associated with an object or a class. It is essentially a function that operates within the context of a class or an instance.
 
  There are two main types of methods in Python:

✲ Instance methods: Operate on an instance of a class and can access its attributes.

✲ Class methods: Operate on the class itself and are defined using @classmethod.

✲ Methods are called on objects (instances of a class) and are passed the instance as the first argument, conventionally named self.

In [9]:
class Welcome:
    def __init__(self, name):  
        self.name = name
    
    def greet(self):
        return f"hello, {self.name}"
        
number = Welcome('world')

print(number.greet()) 

hello, world


Q 2. Explain the concept of function arguments and parameters in python.

In Python, function arguments and parameters are important concepts that define how data is passed to functions.

1. Parameters:

⁍ Parameters are the variables defined in the function signature (the    function definition).

⁍ They act as placeholders for the values that the function will         receive when called.

In [11]:
def curly_tails(story_book): # here (story_book) is a parameter.
    name_ = set(story_book)

    return name_

story_book = ('tails' , 'aladdin' , 'dream girl')
output =  curly_tails(story_book)
print(output)

{'tails', 'dream girl', 'aladdin'}


2. Arguments:

⁍ Arguments are the actual values or data you pass into the function     when you call it.

⁍ These values are assigned to the corresponding parameters in the       function definition.

In [2]:
def curly_tails(story_book):
    name_ = set(story_book)

    return name_

story_book = ('tails' , 'aladdin' , 'dream girl')
output =  curly_tails(story_book)
print(output) #here 'tails' , 'aladdin' , 'dream girl' are the argument.

{'tails', 'aladdin', 'dream girl'}


Q 3.What are the different ways to define and call a function in python?

In Python, there are several ways to define and call functions, depending on the context and type of function. Here are the different approaches:

1. Regular Function Definition and Call

This is the most common way to define and call a function.
Example:

In [4]:
def star_fish(type):
    varity = set(type)
    
    return varity

type = ('dolphin', 'jelly fish' , 'shark')
print(type)

('dolphin', 'jelly fish', 'shark')


2. Function with Default Arguments

You can define functions with default argument values. If the argument is not provided during the call, the default value is used.

Example:

In [15]:
def star_fish(type = 'dolphin'):
    print(type)
  
    
star_fish() #here the default value is present i.e no value is filled in parenthesis

dolphin


3. Function with Keyword Arguments

When calling a function, you can pass arguments by their parameter names, known as keyword arguments. This improves readability and flexibility.

In [12]:
def docu_(personal):
    def small_docu_():
        print(personal)

docu_('aadhar card')

4. Nested Functions

You can define functions inside other functions, known as nested or inner functions.

Example:

In [6]:
def doc_(x , y):
    if (x > y) | (y > -2):
        print('yes, you are right')

    else:
        print ('check again')

x = 89
y = -11
doc_(x , y)

yes, you are right


5.  First-Class Functions (Passing Functions as Arguments)

Functions in Python are first-class citizens, meaning they can be passed as arguments to other functions.

Example:

In [17]:
def element_(x , y):
    return x ** y

output = element_(4 , 3)
print(output)


64


Q 4. What is the purpose of the 'return' statement in a python function?

The return statement in a Python function is used to:

🔸  Exit the function: When the return statement is encountered,      the function terminates, and control is passed back to the caller.

🔸  It can return any data type, such as integers, strings,          lists,tuples, dictionaries, objects, etc.

🔸 If there is no return statement, or if return is used without     an expression, the function returns None by default.

🔸 Once a return statement is executed, the function stops and     no further code inside the function is run.

In [18]:
def story_books(name):
    return (name)

name = 'aaladin'
print(name)

aaladin


In [21]:
def intro_(a , b):
    return (a /b)

a = -110
b = 22
result = a / b
print(result)

-5.0


Q 5. What are operators in python and how do they differ from iterables?


Operators in Python:
Operators in Python are special symbols or keywords that perform operations on values (operands). These operations can be mathematical, logical, comparison-based, or even related to specific types (like strings or lists).

+ : Addition
- : Subtraction
* : Multiplication
/ : Division

Key Differences Between Operators and Iterables:

1. Purpose:
 
▪️  Operators perform actions like addition, comparison, logical operations, and more. They are used to manipulate data.

▪️  Iterables are data structures (like lists, strings, etc.) that contain multiple items and can be looped over or iterated through.

2. Usage:

▪️  Operators are used within expressions, such as 5 + 3, to perform calculations or comparisons.

▪️  Iterables are used to store and access a collection of elements (like a list or string) that you can loop through, like for item in iterable.

3. Data vs Actions:

▪️  Operators define actions (e.g., addition, comparison).

▪️  Iterables are objects containing data that you can traverse.

In [23]:
def maths_(x , y):
    return (x ^ y) # operation is used

x = 179
y = 100
output = maths_( x , y)
print(output)

215


In [24]:
my_list = [11 , 64 , -23 , 80]
for item in my_list:  #iteration used
    print(item)

11
64
-23
80


Q 6.Explain the concept of generators in python and how they are defined.

A generator in Python is a type of iterable, like lists or tuples, but instead of storing all of its values in memory, it generates the values on the fly as you iterate through it. Generators are particularly useful when dealing with large datasets or infinite sequences because they are memory-efficient.

Generators are defined in two main ways:

1. Generator Functions (using yield)
2. Generator Expressions (similar to list comprehensions)

➮  Generator Functions

A generator function is defined like a normal function, but instead of returning a value, it uses the yield keyword. When yield is called, it "pauses" the function and provides the yielded value to the caller. When the function is resumed, it continues from the point where it last paused.

In [25]:
def count_(x):
    while x > -10:
        yield x  # yield will pause the value
        x -= 2

actual_values = count_(10)
for x in actual_values:
    print(x)

10
8
6
4
2
0
-2
-4
-6
-8


➮  Generator Expressions

Generator expressions are similar to list comprehensions, but they produce generators instead of lists. They use parentheses () instead of square brackets [].

In [32]:
gen = (x **x for x in range(4))

for x in gen:
    print(x)

1
1
4
27


Q 7. What are the advantages of using generators over regular funcions ?

These benefits primarily stem from their ability to generate values on the fly without storing everything in memory.

1.Memory Efficiency (Lazy Evaluation):

Generators: Generate values one at a time as needed, without keeping the entire sequence in memory. This allows you to work with large or infinite sequences without running out of memory.

Regular Functions: Typically compute and return all values at once, which can consume a large amount of memory if the dataset is large.

2.  Faster Startup (No Pre-Calculation):

Generators : Do not pre-calculate all values upfront. The values are computed as they are needed, meaning the generator starts almost immediately.

Regular Functions : Might take a long time to start if the entire sequence has to be calculated and returned before the function finishes.

3. Simplified Code for Iteration:
 
 Generators can simplify the process of writing iterators. Instead of manually keeping track of state, a generator function uses yield to automatically "pause" its state and resume it when the next value is requested, making code more concise and easier to maintain.

4. Easier Management of Backpressure:

In scenarios where the producer of data (e.g., an API or a hardware sensor) generates data faster than the consumer can handle it, generators naturally handle backpressure by only producing data when the consumer asks for it.

In [1]:
def funct_():
    for i in range(80):
        yield(i)    

Q 8. What is a function in python and when is it typically used?

A function in Python is a block of organized, reusable code that performs a specific task. Functions help break programs into smaller, modular parts, making them easier to understand, debug, and maintain.    

In [8]:
def calc_sum(x ,y):
    sum = x +y
    print (sum)

x = 9 
y = -0.22
result = calc_sum(x ,y)

calc_sum(-3 , 8.99)

8.78
5.99


## Uses:


Functions are typically used in Python in the following scenarios:

1. Code Reusability:

If you have repetitive code blocks that perform the same task, you can define a function and reuse it instead of writing the same code multiple times.

2. Avoiding Repetition:

Instead of duplicating code for slightly different operations, you can pass different arguments to a function to get varying results while avoiding code duplication.

3. Simplifying Complex Tasks:

You can break down large or complex tasks into smaller, manageable functions, each responsible for a specific subtask. This keeps the code simpler to understand and maintain.

5.  Handling Inputs and Returning Outputs:

Functions can take inputs (parameters) and return outputs, making them flexible and adaptable to different scenarios.
 

In [10]:
def calc_avg(a , b , c):
    sum = a + b + c
    avg = sum / 3
    print(avg)

calc_avg (4 , 8 ,1.5)
calc_avg(-8.9  , 3.2 , 0.11)

4.5
-1.8633333333333333


In [18]:
def calc_complex ( x , y ,z):
    result = x * y - z *(x + z)
    print(result)

calc_complex(2.9 , 6 , -7)

-11.299999999999997


Q 9.Explain the purpose and usage of the 'map( )' function in python.

The map() function in Python is used to apply a given function to all items in an iterable (like a list, tuple, etc.) and return an iterator .
 It is particularly useful when you need to perform the same operation on every element in a collection of data.

In [26]:
def square(y):
    return y * y
   
    
num_ = [ 6 , 2 ,7 ,9 , 5]
sq_num_ = map(square , num_)

print(list(sq_num_))


[36, 4, 49, 81, 25]


In [29]:
def multiply_( x , y ):
    return x* y

list1 = [ 4, 9 , 11, 5 ,20]
list2 = [ 2 , 6 ,10 , 7]

num = map( multiply_ , list1 , list2)
print(list(num))

[8, 54, 110, 35]


Q 10.What is the difference between 'map(  )' , 'reduce(  )' and 'filter(  )' function in python ?

In Python, map(), reduce(), and filter() are higher-order functions used for processing iterables like lists, tuples, etc. They allow for applying transformations or filtering data in a functional programming style.

1, MAP(  )

Purpose: Applies a given function to each item in an iterable (like a list) and returns an iterator (or a map object).

How it works: map() takes a function and one or more iterables. It applies the function to each element of the iterable(s) and returns the result as an iterator. You usually need to convert this iterator to a list or another sequence to view the results.

In [35]:
def division(x , y):
    return( x / y)

list1 = [ 2 , -28 ,108 , 42 , 77 ]
list2 = [ 5 , 9 , 12 , 3 , 11 ]
div_num = map( division , list1 , list2 )

print(list(div_num))

[0.4, -3.111111111111111, 9.0, 14.0, 7.0]


2. FILTER (  )

Purpose: Filters elements from an iterable by applying a function that returns a boolean (either True or False), and only the elements for which the function returns True are included in the result.

How it works: filter() takes a function and an iterable. It applies the function to each element, and if the function returns True, that element is included in the output.

In [39]:
def even_num(x):
    return x % 2 == 0

number = [33 ,16 , 90 ,30 , 33]
result = filter (even_num , number )
print(list(result))

[16, 90, 30]


3. REDUCE (  )

Purpose: Reduces the iterable to a single value by repeatedly applying a function (that takes two arguments) to pairs of elements. It processes the iterable cumulatively, meaning it combines elements step by step into one final result.

How it works: reduce() takes a function that accepts two arguments and applies it cumulatively to the items of the iterable from left to right, reducing the iterable to a single value. It is part of the functools module in Python 3.

In [50]:
from functools import reduce 

def odd_num(x , y):
    return x + y

number = [33 ,16 , 90 ,30 ]
result = reduce(odd_num , number)

print(result)

169


Q 11. Using pen and paper write the internal mechanism for sum operation using reduce function on this given list: [47 , 11 , 42 ,13] ;

In [51]:
from functools import reduce 

def sum_num(x ,y):
    return x + y 

here = [47 , 11 , 42 , 13]
result = reduce(sum_num , here)

print (result)

113


## Steps

Step 1:
 Apply odd_num to the first two elements:

odd_num(47, 11) = 47 + 11 = 58

Step 2:
 Apply odd_num to the result from Step 1 and the next element:

odd_num(58, 42) = 58 + 42 = 100

Step 3:  
 Apply odd_num to the result from Step 2 and the final element:

odd_num(100, 13) = 100 + 13 = 113

So, the output of reduce(odd_num, [47, 11, 42, 13]) is 113

## PRACTICE Q.

Q 1. Write a python function that takes a list of numbers as input and returns the sum of all even nos. in  the list .

In [3]:
from functools import reduce  

def only_even(a , b):
    if b % 2 ==0:
        return a + b
        
    else:
         return a 

list1 = [2 , 7.1 , 17 , 0.2 , 4] 
result = reduce( only_even , list1)

print(result) # float values are not operated by reduce()

6


Q 2. Create a python function that accepts a string and the reverse of that string. 

In [28]:
def rev_strin(input_string):
    return input_string [ : :-1]
  
input_string =  'I am the boss'
reverse_string = rev_strin(input_string)

print("input_string:" , input_string)
print( " reverse_string:" , reverse_string)

input_string: I am the boss
 reverse_string: ssob eht ma I


Q 3. Implement a python function that takes a list of integers and return a new list containing the square of each no. ?   

In [ ]:
def square_lis(x):
    return x**2

x = [3 , 8 ,10 ,9 ,14]
final = list (map(square_lis , x))

print(final)

Q 4. Write a python function that checks if a given no. is prime or not from 1 to 200.

In [27]:
  def prime_num( h ):
    if h < 2:
        return False

    for i in range ( 2 , int (h**0.5) + 1):
        
        if h % i == 0:
            return False
    
    return True       
    
for h in range( 1 , 201 ):
     if prime_num(h):
            print(h , end = ' , ')
            

2 , 3 , 5 , 7 , 11 , 13 , 17 , 19 , 23 , 29 , 31 , 37 , 41 , 43 , 47 , 53 , 59 , 61 , 67 , 71 , 73 , 79 , 83 , 89 , 97 , 101 , 103 , 107 , 109 , 113 , 127 , 131 , 137 , 139 , 149 , 151 , 157 , 163 , 167 , 173 , 179 , 181 , 191 , 193 , 197 , 199 , 

Q 5. Create on iterator class in python that generates the fibonacci sequence up to a specified no. of terms.

Q 6. Write a generator function in python that yields the powers of 2 up to a given exponents. 

In [40]:
def expo_num(value):
    
    for exp in range(value):
        yield 2**exp
    
for number in expo_num(6):
    print(number )

1
2
4
8
16
32


Q 7. Implement a generator function that reads a file line by line and yields each line as a string.

In [ ]:
def mys_string(file_path):
    
    with open(file_path, 'r') as file:
        for line in file:
            yield line.strip()
        
file_path = 'hello world'  

for line in mys_string(file_path):
    print(line)        

Q 8.Use a lamba function in python to sort a list of tuples based on the second siement of each tuple.

In [50]:
tuples__ = [(1 , 'pink') , (2 , 'yellow') , (3 , 'blue') , (4 , 'grey')]

new_list = sorted(tuples__ , key = lambda x : x [1])

print(new_list)

[(3, 'blue'), (4, 'grey'), (1, 'pink'), (2, 'yellow')]


Q 9. Write a python program that use 'map( )' to convert a list to temperature from celsius to fahrenheit.

In [1]:
def celsius_to_fahrenheit(celsius):
    return(celsius * 9/5) + 32

celsius = [ 80 , 62 , 27.5 , 39.50]
fahrenheit_list = list(map( celsius_to_fahrenheit , celsius))

print(fahrenheit_list)

[176.0, 143.6, 81.5, 103.1]


Q 10. Create a python program that uses 'filter(  )' to remove all the vowels from a given string.

In [ ]:
def not_vowel(alpha):
    vowels = 'aeiouAEIOU'
    return alpha not 

In [8]:
def not_vowel(alpha):
    vowels = 'aeiouAEIOU'
    return alpha not in vowels

string1 = 'This is a simple book.'

result = ''.join(filter(not_vowel , string1))

print(result)

Ths s  smpl bk.


"Good morning, sir. I have some queries on certain topics, which is why I wasn't able to solve those questions. Kindly help me with my queries."